### PR TITLE
[TE] Integrate TE router 

### PR DESCRIPTION
**Pull request by tdophung**
# Description
TE's router has a potential to be faster than the pure JAX router currently used in maxtext. Especially when coupled with TE's permutation (with `use_te_permutation` config), permutation will directly consume the output from router in the form of a binary routing map, without needing to adapt the output from a list of indices. 

More performance data can be uploaded later

*Notice 1:* Once all tests pass, the "pull ready" label will automatically be assigned.
This label is used for administrative purposes. Please do not add it manually.

*Notice 2:* For external contributions, our settings currently require an approval from a MaxText maintainer to trigger CI tests.

# Tests
I compared the loss in 30 steps from training mixtral-8x7b.yaml (with only 4 layers) between MaxText's pure jax code path with the loss when using te_router and they are the same.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [ ] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [ ] I have necessary comments in my code, particularly in hard-to-understand areas.
- [ ] I have run end-to-end tests tests and provided workload links above if applicable.
- [ ] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -2382,6 +2386,14 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
         raise ValueError("GPT-OSS MoE only supports dropless (capacity_factor=-1) with dense matmul.")
       if self.routed_bias and self.routed_bias_update_rate > 0.0 and self.decoder_block != DecoderBlockType.DEEPSEEK:
         raise ValueError("Loss-free load balancing is only supported for the DeepSeek decoder block.")
+      if self.te_router_impl:
+        if not self.te_permutation_impl:
```
Comment (jberchtold-nvidia) [inline src/maxtext/configs/types.py]: Since TE router+permute are coupled, would it make sense to make these a single config flag like `te_use_router_and_permute` which would let us remove the additional complexity of supporting permute alone and having to convert mask types from the pure-JAX router?

File: src/maxtext/layers/moe.py [hunk 2]
```diff
@@ -648,43 +658,44 @@ def apply_ffn_activation(self, layer_w0, layer_w1):
...
     if roll_to_expert_id is not None:
       flatten_selected_experts = (flatten_selected_experts - roll_to_expert_id) % self.num_experts
     sorted_selected_experts = jnp.argsort(flatten_selected_experts)
-    # sort inputs for number of selected experts
     replicated_inputs_2d = jnp.repeat(inputs_2d, self.num_experts_per_tok, axis=0)
     sorted_inputs = _sort_activations(replicated_inputs_2d, sorted_selected_experts, use_custom_sort_vjp).astype(
         self.dtype
     )
     group_size = jnp.bincount(flatten_selected_experts, length=self.num_experts)
-    # Return the experts for each sorted input.
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py:687]: Were these comments intended to be removed?

File: src/maxtext/configs/models/mixtral-8x7b.yml [hunk 3]
```diff
@@ -20,7 +20,7 @@ base_num_query_heads: 32
 base_num_kv_heads: 8
 base_mlp_dim: 14336
 base_moe_mlp_dim: 14336
-base_num_decoder_layers: 32
```
Comment (jberchtold-nvidia) [inline src/maxtext/configs/models/mixtral-8x7b.yml:23]: Self-resolve: Don't commit this local change

File: src/maxtext/configs/base.yml [hunk 4]
```diff
@@ -189,7 +189,7 @@ load_balance_loss_weight: 0.0 # weight for the load balance loss
 use_random_routing: false # whether to use random routing for debug/test purpose
 use_custom_sort_vjp: true # whether to use a custom VJP sort for efficient backward pass processing in sparse matmul
 use_ring_of_experts: false # whether to use ring of experts for sparse matmul expert parallelism
-te_permutation_impl: false # whether to use TransformerEngine permutation kernels for MoE token dispatch/combine
+te_permutation_impl: false # whether to use TransformerEngine fused router and permutation kernels for MoE routing and token dispatch/combine
```
Comment (jberchtold-nvidia) [inline src/maxtext/configs/base.yml]: Self-resolve: Rename to `te_router_and_permutation_impl` or something similar. This still won't be the official name, we'll have time to reassess the best name when we officially upstream, but would be good to include router in the temporary name

File: src/maxtext/layers/moe.py [hunk 5]
```diff
@@ -928,25 +937,25 @@ def permute(self, inputs, gate_logits, pre_bias_logits, use_custom_sort_vjp=True
       - lb_loss: Load balance loss (or None).
       - bias_updates: Bias updates (or None).
     """
-    use_te = getattr(self.config, "te_permutation_impl", False) and te_permutation.TE_PERMUTATION_AVAILABLE
+    use_te = getattr(self.config, "te_permutation_impl", False) and te_permutation.TE_PERMUTATION_AVAILABLE and te_router.TE_ROUTER_AVAILABLE
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: Instead of the silent fallback to MaxText route+permute here, can we raise an error if te route and permute is enabled and either TE_PERMUTATION_AVAILABLE or TE_ROUTER_AVAILABLE is false.

Previously, these two routes+permutes were functionally equivalent. However, now with the padding/alignment of groups, TE's route+permute has more features so falling back to MaxText and ignoring group alignment isn't functionally equivalent and will cause issues for downstream items that require group alignment (e.g. grouped GEMM)

File: src/maxtext/layers/moe.py [hunk 6]
```diff
@@ -1374,6 +1383,14 @@ def sparse_matmul(
   ):
     """Perform sparse matrix multiplication of inputs and Experts."""
 
+    # Extract gate expert bias internally when TE path is active.
+    # This avoids exposing gate_expert_bias in the public API.
+    use_te = getattr(self.config, "te_permutation_impl", False) and te_permutation.TE_PERMUTATION_AVAILABLE and te_router.TE_ROUTER_AVAILABLE
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: Same comment here about silent fallback. Maybe the best option is to make the functions inside te_permutation and te_router throw if TE_PERMUTATION_AVAILABLE or TE_ROUTER_AVAILABLE is false, that way we don't have to remember to check it at all usages in moe.py

File: src/maxtext/layers/moe.py [hunk 7]
```diff
@@ -2505,7 +2526,13 @@ def __call__(
   ) -> tuple[jax.Array, Optional[jax.Array], Optional[jax.Array]]:
     cfg = self.config
     inputs = inputs.astype(cfg.dtype)
-    gate_logits, pre_bias_logits = self.gate(inputs)
+
+    use_te = getattr(cfg, "te_permutation_impl", False) and te_permutation.TE_PERMUTATION_AVAILABLE and te_router.TE_ROUTER_AVAILABLE
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: Same here about silent fallback

File: src/maxtext/layers/moe.py (hunk 6 above)
Comment (jberchtold-nvidia) [reply src/maxtext/layers/moe.py]: Seems like the erroring may already be handled in te_router.py via `check_te_router_available()` calls. Is this also true for te_permutation.py?

If so, all that will be needed here is to remove ` and te_permutation.TE_PERMUTATION_AVAILABLE and te_router.TE_ROUTER_AVAILABLE`

File: src/maxtext/layers/te_router.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+  Args:
+    probs: Score tensor from te_compute_aux_scores [num_tokens, num_experts].
+    tokens_per_expert: Token counts per expert [num_experts]. Integer tensor.
+    topk: Top-k value.
+    coeff: Loss coefficient (e.g., load_balance_loss_weight).
+
+  Returns:
+    Scalar loss value.
+  """
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/te_router.py:202]: Can we remove any complexity from `te_permutation.py` now that the router is integrated?

Such as, `create_dense_probs_from_topk`, `create_routing_map_from_indices` which I believe were only for converting the routing map from the MaxText router, which we no longer need now that the TE router was integrated. The usages of these functions seems to be removed in this PR

Also, are `compute_ragged_all_to_all_params` and `compute_reverse_ragged_all_to_all_params` still needed? Seems like the answer is yes as I still see them used. The reason we still need these is because it's required for adapting TE permute outputs to ragged A2A, and having the TE router doesn't change this because the router runs before TE permute and doesn't affect the data between TE permute and ragged A2A, right?

File: src/maxtext/layers/moe.py [hunk 9]
```diff
@@ -705,125 +729,109 @@ def _te_permute(
...
+    num_out_tokens = num_tokens * self.num_experts_per_tok
 
-    # Get top-k routing decisions
-    # Note: weights are already softmaxed over the k selected experts in get_topk()
-    weights, top_k_indices = self.get_topk(gate_logits, pre_bias_logits, rngs)
+    # --- TE Router ---
+    raw_logits_2d = gate_logits.reshape(num_tokens, -1)
+
+    router_dtype = jnp.float32 if self.config.logits_dot_in_fp32 else self.dtype
+    raw_logits_2d = raw_logits_2d.astype(router_dtype)
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: Since the `_te_permute` function is getting more complex with this PR, can we move it to `te_router.py` or `te_permutation.py` to minimize our diff to upstream MaxText for rebasing.

`_te_unpermute` wasn't changed, but might as well move it too if we move `_te_permute`

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (jberchtold-nvidia) [reply src/maxtext/configs/types.py]: Outdated from previous review cycle that was discussed offline. Resolving

File: src/maxtext/layers/te_router.py (hunk 8 above)
Comment (tdophung) [reply src/maxtext/layers/te_router.py:202]: Yes the complication for the router such as `create_dense_probs_from_topk` amnd `create_routing_map_from_indices` can be removed

No, the `compute_ragged_all_to_all_params` and `compute_reverse_ragged_all_to_all_params` cannot be removed as these are adapters to ragged A2A, nothing to do with router. The maxtext path also have their compute_ragged_all_to_all_params function to adapt but ours is slightly different.

Comment (jberchtold-nvidia) [reply src/maxtext/layers/te_router.py:202]: Sounds good, thanks!

File: src/maxtext/layers/moe.py (hunk 2 above)
Comment (tdophung) [reply src/maxtext/layers/moe.py:687]: No I added them back

File: src/maxtext/layers/moe.py [hunk 10]
```diff
@@ -1374,6 +1263,14 @@ def sparse_matmul(
   ):
     """Perform sparse matrix multiplication of inputs and Experts."""
 
+    # Extract gate expert bias internally when TE path is active.
+    # This avoids exposing gate_expert_bias in the public API.
+    use_te = getattr(self.config, "te_router_and_permutation_impl", False)
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: Can we do
```
use_te = self.config.te_router_and_permutation_impl
```
instead of the `getattr` way?

I often have multiple MaxText versions checked out for various tests and it's easy to point to the wrong base.yml by mistake. With the `getattr` way, this will be a more subtle error as it will default to `False` here. With the `self.config.te_router_and_permutation_impl` way, it'll be clear that `te_router_and_permutation_impl` doesn't exist on the config and that I've used the wrong config by mistake.

Lmk if there's a use case for `getattr` that I'm missing, thanks!

File: src/maxtext/layers/moe.py [hunk 11]
```diff
@@ -2505,7 +2406,13 @@ def __call__(
   ) -> tuple[jax.Array, Optional[jax.Array], Optional[jax.Array]]:
     cfg = self.config
     inputs = inputs.astype(cfg.dtype)
-    gate_logits, pre_bias_logits = self.gate(inputs)
+
+    use_te = getattr(cfg, "te_router_and_permutation_impl", False)
```
Comment (jberchtold-nvidia) [inline src/maxtext/layers/moe.py]: Same getattr question as above

File: src/maxtext/layers/moe.py (hunk 10 above)
Comment (tdophung) [reply src/maxtext/layers/moe.py]: I thought it would be a gentler way of handling users that do not have the maxtext version with te_router_and_permutation_impl using the same script for version that have it. But I guess this is the same way of thinking that led to the fall back to MT's impl if we do not have triton. So getting rid of `getattr()` would be clearer?

Comment (jberchtold-nvidia) [reply src/maxtext/layers/moe.py]: Yeah I think it's not a clear decision either way, I see your point for the other case of smoothing the transition. Given we have lots of moving pieces and knobs for users to test and this te/main MaxText branch is experimental, my preference is to lean towards overly cautious errors instead of backwards compatibility for now. That will help us and early adopters debug issues easier as we scale this up. But when things are a bit more stabilized and we're ready to upstream this, I'm not opposed to switching back as I do agree your version with `getattr` is better for compatibility if users are using different versions that aren't in experimental stages.

**Review by jberchtold-nvidia — APPROVED** [review]
